baybe/utils/augmentation.py

```diff
@@ -0,0 +1,220 @@
+) -> pd.DataFrame:
+    """Augment a dataframe if permutation invariant columns are present.
+
+    *   Original
+
+        +----+----+----+----+
+        | A1 | A2 | B1 | B2 |
+        +====+====+====+====+
+        | a  | b  | x  | y  |
+        +----+----+----+----+
+        | b  | a  | x  | z  |
+        +----+----+----+----+
+
+    *   Result with ``columns = [["A1"], ["A2"]]``
+
+        +----+----+----+----+
+        | A1 | A2 | B1 | B2 |
+        +====+====+====+====+
+        | a  | b  | x  | y  |
+        +----+----+----+----+
+        | b  | a  | x  | z  |
+        +----+----+----+----+
+        | b  | a  | x  | y  |
+        +----+----+----+----+
+        | a  | b  | x  | z  |
+        +----+----+----+----+
+
+    *   Result with ``columns = [["A1", "B1"], ["A2", "B2"]]``
+
+        +----+----+----+----+
+        | A1 | A2 | B1 | B2 |
+        +====+====+====+====+
+        | a  | b  | x  | y  |
+        +----+----+----+----+
+        | b  | a  | x  | z  |
+        +----+----+----+----+
+        | b  | a  | y  | x  |
+        +----+----+----+----+
+        | a  | b  | z  | x  |
+        +----+----+----+----+
+
+    Args:
+        df: The dataframe that should be augmented.
+        columns: Sequences of permutation invariant columns. The n'th column in each
+            sequence will be permuted together with each n'th column in the other
+            sequences.
+
+    Returns:
+        The augmented dataframe containing the original one.
+
+    Raises:
+        ValueError: If ``dependents`` has length incompatible with ``columns``.
+        ValueError: If entries in ``dependents`` are not of same length.
+    """
+    # Validation
+    if len(columns) < 2:
+        raise ValueError(
+            "When augmenting permutation invariance, at least two column sequences "
+            "must be given."
+        )
+    if len({len(seq) for seq in columns}) != 1 or len(columns[0]) < 1:
+        raise ValueError(
+            "Permutation augmentation can only work if the amount of columns un each "
+            "sequence is the same and the sequences are not empty."
+        )
+
+    # Augmentation Loop
+    new_rows: list[pd.DataFrame] = []
+    idx_permutation = list(permutations(range(len(columns))))
+    for _, row in df.iterrows():
+        to_add = []
+        for _, perm in enumerate(idx_permutation):
+            new_row = row.copy()
+
+            # Permute columns
+            for deps in map(list, zip(*columns)):
+                new_row[deps] = row[[deps[k] for k in perm]]
+
+            # Check whether the new row is an existing permutation
+            if not _row_in_df(new_row, df):
+                to_add.append(new_row)
+
+        new_rows.append(pd.DataFrame(to_add))
+    augmented_df = pd.concat([df] + new_rows)
+
+    return augmented_df
+
+
+def df_apply_dependency_augmentation(
+    df: pd.DataFrame,
+    causing: tuple[str, Sequence],
+    affected: Sequence[tuple[str, Sequence]],
```

Comment:
I'll let you double-check, but must the outer type really be a `Sequence` or couldn't it just be a `Collection` instead?

Reply:
what are your arguments? the inner sequences in my view are still homogeneous types, so a Collection implying heterogeneous elements doesnt seem right? But I always forget the differences between these macrotypes...

Reply:
Has nothing to to with homogeneous / heterogeneous but what kind of methods/properties you require to exist. `Sequence` would be only required if you need the inner elements to be ordered and have a finite length. I think – and please correct me if I'm wrong – that the ordering is not required here!?

Reply:
(taling abou the outer type hint)
so if it needs to be finite it should be `Sequence` ?

Reply:
Then I think `Collection` is the right type 👍🏼 To answer the question: No, finite does not imply `Sequence` (e.g. `Collection` and `Set` is also finite), but `Sequence` -> finite does hold.

Reply:
so I updated type hints, for the ermutation function it still needs to be `Sequence` because for valdiaiton I need to inde it
for dependency augmentation I made it `Iterable` sicne its only sued with `*` oeprator

Reply:
I don't think `Iterable` would work, though. I suspect you cannot handle an infinite stream of data? :D

Reply:
ok then let me know whats the best type for finite-iterable (and dont say its `Sequence`)

Reply:
It's `Sequence` of course .... just kidding 🤣 

I have in fact already mentioned it twice, [here](https://github.com/emdgroup/baybe/pull/290#discussion_r1664005385) and [here](https://github.com/emdgroup/baybe/pull/290#discussion_r1670392981): I think the right type should be `Collection` (it is finite, iterable, but order doesn't matter)

Reply:
changed